doc/man1/fts_harvester.1
.\" Man page for fts_harvester
.\" Use the following command to view man page:
.\"
.\"  tbl fts_harvester.1 | nroff -man | less
.\"
.TH FTS_HARVESTER 1 "March 26, 2015"
.SH NAME
fts_harvester \- A tool to harvest Emdros databases with Full Text Search
.SH SYNOPSIS
\fBfts_harvester\fR [ options ] <query-terms...>
.br
\fBfts_harvester\fR <--help|-V|--version>
.br
.SH DESCRIPTION
\fBfts_harvester\fR is a command-line tool to harvest an Emdros
database that has been prepared for Full Text Search by fts_indexer(1)
or via the libharvest library.

The \fBfts_harvester\fR tool needs a source database to read from, and
at least one query term.

The program also needs a "bookcase object type", which must consist of
objects that are somehow "larger" than the "indexed object type". (In
Information Retrieval terms, the "bookcase_object_type" is the
"document".)  That is, the "indexed object type" (for example, "Word")
must have objects which are wholly contained within the "bookcase
object type" (for example, "document").  In effect, the "bookcase
object type" provides the frames of monads within which to index the
"indexed object type". In addition, the "indexed object type" must
have a concomitant "indexed feature", which is the feature on which to
index.  For example, "surface" could be a feature on "Word" which
could be ideal to index on.

The output of the program is either plain text or XHTML snippets (if
the --xhtml option is given).

The program can optionally show titles, taken from a "bookcase title
object type" having a "bookcase title feature".  If, for example,
there is an object type called "part", having a feature "part_name",
then the former can be given as the bookcase title object type (using
the option "--bookcase-title-otn"), and the feature "part" can be
given as the boookcase title feature (using the option
"--bookcase-title-feature").


.SH OPTIONS
\fBfts_harvester\fR supports the following command-line switches:
.TP 6
.BI \-\-help
show help
.TP
.BI \-V\ ,\ \-\-version
show version
.TP
.BI \-fe\ ,\ \-\-fts\-engine \ fts\-engine\-version
Use the given version of the FTS engine. Currently, versions 1 and 2
are supported. Version 1 is the current default.
.TP
.BI \-g\ ,\ \-\-google-syntax
If given, each query term, if surrounded by "double quotes", is
interpreted as a series of terms which must appear in (monad) sequence
right after each other.  Example: \fB-g '"the prime minister"'\fR.
.TP
.BI \-\-xhtml
If given, XHTML snippets are emitted.  The XHTML preamble and the
ending </body></html> snippet are not emitted.
.TP
.BI \-d\ ,\ \-\-dbname \ dbname 
set database from which to read, and to which to write (unless -td or
--nodb is given).
.TP
.BI \-lp\ ,\ \-\-largest-proximity \ integer
Use this number as the largest proximity (in monads) to allow between
query terms (within the boundaries of each object in the bookcase
object type).
.TP
.BI \-\-bookcase\-otn \ object\-type\-name
Use object-type-name as the object type which surrounds the "indexed
object type".
.TP
.BI \-\-indexed\-otn \ object\-type\-name
Use object-type-name as the object type which must be indexed.
.TP
.BI \-\-indexed\-feature \ feature\-name
Use feature-name as the feature on the indexed object type for which
to generate the index.  That is, if the indexed object type is, for
example, "Word", then the indexed feature could be "surface", and then
"Word" must have a feature called "surface".
.TP
.BI \-\-bookcase\-title\-otn \ object\-type\-name
Use object-type-name as the object type from which to get a title for
each hit.
.BI \-btf\ ,\ \-\-bookcase\-title\-feature \ feature\-name
Use feature-name as the feature on the bookcase title object type from
which to get a title for each hit.  For example, if there is an object
type callled "part", with a feature called "name", then "name" might
be given here, and "part" might be given for --bookcase-title-otn.
.TP
.BI \-o\ ,\ \-\-output \ filename
dump MQL statements to file filename. If this switch is not given,
then the MQL is just executed within the target database.
.TP
.BI \-h\ ,\ \-\-host \ hostname 
set source db back-end hostname to connect to (default
is 'localhost').  Has no effect on SQLite or SQLite 3.  If --nodb is
not given, and if -th or --target-host is not given, then the target
database host will be the same as the source database host.
.TP
.BI \-u\ ,\ \-\-user \ user
set source database user to connect as (default is 'emdf').  Has no
effect on SQLite or SQLite 3.  Will be used for the target database
user as well, unless --nodb, -tu, or --target-user is given.
.TP
.BI \-p\ ,\ \-\-password \ password
set source database password to use for the source database user.  Has
no effect on SQLite or SQLite 3, unless you have an encryption-enabled
SQLite (3), in which case this gets passed as the key.  Will be used
for the target database password, unless one of --nodb, -tp or
--target-password is given.
.TP
.BI \-b\ ,\ \-\-backend \ backend
set source database backend to `backend'. The target database backend
is also given with this switch, unless provided by the -tb switch.
Valid values are: For PostgreSQL: "p", "pg", "postgres", and
"postgresql". For MySQL: "m", "my", and "mysql". For SQLite 2.X.X:
"2", "s", "l", "lt", "sqlite", and "sqlite2". For SQLite 3.X.X: "3",
"s3", "lt3", and "sqlite3".
.TP
.BI \-page\ ,\ \-\-page \ page-number
In XHTML mode, gives the page to start at (first page == 1). If < 1,
all results are returned. Defaults to -1, so all results are displayed
by default.
.TP
.BI \-hpp\ ,\ \-\-hits-per-page \ max-number-of-hits-per-page
Sets the maximum number of hits per page. Defaults to 10.
.TP
.BI \-sf\ ,\ \-\-stylesheet-filename \ stylesheet-filename
Gives the name of the file containing the JSON stylesheet to use. See
the man-page fts_filters(5) to learn the syntax of that file.

This option must be used in conjunction with the -s option.  If it is
used, you cannot also use filters at the end of the command line.
.TP
.BI \-s\ ,\ \-\-stylesheet \ stylesheet-name
This option tells the program which stylesheet within the stylesheet
file provided with the -sf option must be used.

This option must be used in conjunction with the -sf option.

.SH FILTERS
See fts_filters(5) for a discussion of filters.  See the options -sf
and -s for how to use them.

.SH RETURN VALUES
.TP
.BR 0 " Success"
.TP
.BR 1 " Wrong usage"
.TP
.BR 2 " Connection to backend server could not be established"
.TP
.BR 3 " An exception occurred (the type is printed on stderr)"
.TP
.BR 4 " Could not open file"
.TP
.BR 5 " Database error"
.TP
.BR 6 " Compiler error (internal error)
.SH AUTHORS
Written Ulrik Sandborg-Petersen (ulrikp@emdros.org).


NEWS
This file is for release-notes for Emdros.

- *** Version 3.7.0 *** 

Release notes for Emdros 3.7.0
==============================

- Emdros has been relicensed under the MIT license.

- All copyright in Emdros and related documentation and build scripts
  has been assigned to Ulrik Sandborg-Petersen's holding company,
  Sandborg-Petersen Holding ApS.  This is so as to be able to serve
  customers better with support and other Emdros-related help.

- A new construct was added to the topographic part of MQL: computed
  features, which take a monad set and returns a monad or integer:

  - first_monad(<monad_set_name>) returns the smallest (first) monad.
  - last_monad(<monad_set_name>) returns the largest (last) monad.
  - monad_count(<monad_set_mame>) returns the cardinality.
  - monad_set_length(<monad_set_mame>) returns last_monad - first_monad + 1.

  If the monad_set_name (and the parentheses) are missing, then the
  privileged monad set "monads" is assumed.

  For example:

  SELECT ALL OBJECTS
  WHERE
  [Clause
     monad_set_length > 15 AND monad_count <= 15
  ]

  This also works with object references:

  // Finds sentences and clauses which are co-extensive.
  SELECT ALL OBJECTS
  WHERE
  [Sentence AS s1
     [Clause monad_count = s1.monad_count]
  ]

- The following SWIG bindings have been removed:

  - PHP 5
  - Ruby
  - Perl
  - Android

  If any user needs one of these back, please let Ulrik know.

  These SWIG bindings were removed because:

  - The language was EOL'ed (PHP5)
  - The Emdros SWIG bindings had not been tested for several years (Ruby, Perl)
  - They were redundant (Android).

  Users of Android should simply use the Java SWIG bindings instead.

- The Emdros Query Tool no longer supports TECkit.  TECkit support was
  removed because we now live in a Unicode world, and TECkit was
  supposed to convert from legacy 8-bit encodings to Unicode.

- The examples/ subdirectory is gone, making for a smaller tarball.
  The examples have been moved to the example-programs-emdros and
  example-databases GitHub repos. This affects:

  - The Emdros Chunking Tool
  - The HAL implementation (Hyperspace Analogue to Language of Burgess et al.)
  - The King James Version, Parsed.
  - The sfm example.

- The agexport(1) program is gone, since it had not been maintained or
  tested for many years.

- The build scripts have been totally reimagined.  In particular:

  - The m4/ directory has been totally reorganized.

  - The doc/ subdirectory has been totally revamped, with much
    documentation having been moved to the docs-emdros repo on GitHub.

  - The ./configure script now takes a mutually exclusive pair of
    switches: --enable-devel | --disable-devel switch.

    The default is --disable-devel.

    If --disable-devel is true, then the Emdros libraries,
    libpcre_emdros.*, libemdf.*, libmql.*, libutil.*, and
    libharvest.*.  Instead, the amalgamation is used in lieu of these.
    In addition, these libraries are not compiled, thus providing a
    much faster compile experience.  In addition, the Emdros programs
    are linked against the amalgamation library.

    If --enable-devel is true, then everything is compiled, including
    the amalgamation, but the programs are linked against the
    development libraries, not the amalgamation.  This is useful if
    you are developing Emdros.


- *** Version 3.6.1 *** 

Release notes for Emdros 3.6.1
==============================

Emdros version 3.6.1 is a maintenance release with the following
summary of changes since 3.6.0:

- The MQL query language has been extended with the ability to compare
  features with type LIST OF INTEGER, LIST OF ID_D, and LIST OF
  <enum-type> with an explicit list (including the empty list) using
  the "=" comparison operator.

  For example:

  SELECT ALL OBJECTS
  WHERE
  [Clause_atom
     daughters = () // Test for empty list
     OR
     daughters = (12345,67891) // Test for explicit, non-empty list
  ]
  GO

- Detection of PHP 5 and PHP 7 on macOS now works correctly.

- Compileability on Solaris 10 with Sun's CC C++ compiler has been
  restored.

- Compatibility with wxWidget 2.8.12 has been restored.


Enjoy!


Ulrik Sandborg-Petersen



- *** Version 3.6.0 *** 

Release notes for Emdros 3.6.0
==============================

The release notes below cover some of the important changes in Emdros
since 3.5.0.

- SQLite2 support has been removed, as version 2 of SQLite has not
  been updated by the SQLite authors for many, many years.  Also,
  SQLite2 support in Emdros has been been deprecated for some time.

- The MQL subsystem has a better query planner, which is able to
  choose the "Innermost First" strategy, rather than the default
  "Outermost First" strategy, under many more circumstances.

- The Emdros MySQL backend now has support for transactions, if
  available in the MySQL server.

- Assorted performance improvements to the following subsystems:

  - XML parsing (renderobjects, for example)
  - MQL

- Node.JS support has been removed.  The reason is that the latest
  SWIG version (3.0.12) does not support the latest Node.  Also, the
  Node.JS bindings never really worked.

- PCRE has been upgraded.

- SQLite3 has been upgraded.

- JSON output now escapes string characters correctly.

- Compiling Linux .rpm packages works on recent systems (CentOS 7,
  Fedora 28).

- Compiling Linux .deb packages works on recent systems (Ubuntu 16.04,
  Ubuntu 18.04, Debian 9.4).

- Building on Windows supports the latest Visual Studio.


src/fts_harvester.cpp
/*
 * fts_harvest.cpp
 *
 * A front-end to the BookcaseHarvester class in the libharvest library.
 * Created: 5/1-2001 (1st of May, 2001)
 * Last update: 5/11-2018
 *
 * Return codes:
 * 
 * 0 : Success
 * 1 : Wrong usage
 * 2 : Connection to backend server could not be established
 * 3 : An exception occurred
 * 4 : Could not open file
 * 5 : Database error
 * 6 : Compiler error
 *
 */

/*
 * Copyright (C) 2001-2018     Ulrik Sandborg-Petersen
 * Copyright (C) 2018-present  Sandborg-Petersen Holding ApS, Denmark
 *
 * Licensed under the MIT License.
 *
 * Please see the file COPYING in the root of the sources for more details.
 *
 */



#include <emdros-config.h>
// Do this because emdros-config might #define malloc,
// in which case some systems will fail to use the real 
// malloc, thus causing build errors.
#undef malloc


#include <monads.h>
#include <iostream>
#include <sstream>
#include <fstream>
#include <string_func.h>
#include <emdfdb.h>
#include <table.h>
#include <emdros-lconfig.h>
#include <emdf_output.h>
#include <emdros_environment.h>
#include <encryption.h>
#include <opt.h>
#include <messages.h>

#include <harvest_fts.h>
#include <harvest_fts3.h>


int exec_harvester(EmdrosEnv *pEnv, const std::string& output_filename, int fts_engine_version, const std::string& bookcase_otn, const std::string& indexed_otn, const std::string& indexed_feature, const std::string& stylesheet_name, const std::string& stylesheet_file_name, const StringList& query_string_list, long largest_proximity, bool bUseGoogleSyntax, bool bGetXHTML, bool bSingleBookcaseHits, const std::string& bookcase_title_otn, const std::string& bookcase_title_feature, int hits_per_page, int page_number)
{ 
	bool bResult = true;

	std::ofstream *pFOUT = 0;
	if (output_filename.empty()) {
		pFOUT = 0;
	} else {
		pFOUT = new std::ofstream(output_filename.c_str());
	}

	if (pFOUT != 0 && !(*pFOUT)) {
		delete pFOUT;
		std::cerr << "ERROR: Could not open file '" << output_filename << "' for writing." << std::endl;
		return 4;
	}

	// from harvest/harvest_fts.cpp.
	EmdrosFTSStringFilter *pFilterChain = 0;

	if (stylesheet_name == "") {
		pFilterChain = 0;
	} else {
		std::string filter_error_message;
		pFilterChain = stringFilterFactoryJSONFileName(stylesheet_file_name, bResult, filter_error_message, stylesheet_name);
		if (!bResult) {
			std::cerr << "ERROR doing parsing the filters:\n" 
				  << filter_error_message
				  << std::endl;
		}
	}

	if (bResult) {
		BookcaseHitList result;
		std::string error_message;
			
		std::string query;
		StringListConstIterator qlci = query_string_list.const_iterator();
		while (qlci.hasNext()) {
			std::string this_term = qlci.next();
			query += this_term;
			if (qlci.hasNext()) {
				query += " ";
			}
		}
			
		if (fts_engine_version == 3) {
			bResult = harvestFTS3(pEnv, bUseGoogleSyntax, bookcase_otn, indexed_otn, indexed_feature, SetOfMonads(1, MAX_MONAD), query, largest_proximity, bSingleBookcaseHits, error_message, result);
		} else {
			if (bUseGoogleSyntax) {
				bResult = harvestFTSGoogle_with_version(pEnv, fts_engine_version, bookcase_otn, indexed_otn, indexed_feature, SetOfMonads(1, MAX_MONAD), query, largest_proximity, pFilterChain, bSingleBookcaseHits, error_message, result);
			} else {
				bResult = harvestFTSNonGoogle_with_version(pEnv, fts_engine_version, bookcase_otn, indexed_otn, indexed_feature, SetOfMonads(1, MAX_MONAD), query, largest_proximity, pFilterChain, bSingleBookcaseHits, error_message, result);
			}
		}
			
		if (bResult) {
			std::ostream *pOut;
			if (pFOUT == 0) {
				pOut = &std::cout;
			} else {
				pOut = pFOUT;
			}
			if (bGetXHTML) {
				// FIXME: Replace indexed_otn with token_otn
				// FIXME: Replaced indexed_feature with token_feature_name
				// FIXME: Replace 8 and 12 with monads_before_hit_word and monads_after_hit_word respectively
				result.setMaxHitsPerPage(hits_per_page);
				bResult = result.harvestHits(pEnv, bookcase_otn, bookcase_title_otn, bookcase_title_feature, indexed_otn, indexed_feature, 8, 12, query, pFilterChain, page_number);
				if (bResult) {
					BookcaseHitListConstIterator my_ci = result.const_iterator(page_number);
					if (my_ci.hasNext()) {
						(*pOut) << "<dl>\n";
						while (my_ci.hasNext()) {
							BookcaseHit myhit(my_ci.next());
							std::string snippet;
							std::string snippet_title;
							result.getSnippet(myhit, "<b>", "</b>", " ", true, snippet, snippet_title);
							(*pOut) << "<dt>" << snippet_title << "</dt>\n";
							(*pOut) << "<dd>" << snippet << "</dd>\n";
						}
						(*pOut) << "</dl>\n";
					}
				}
				if (page_number > 0) {
					(*pOut) << "<p>" << "Showing page " << page_number << " of " << result.getPageCount() << " pages.</p>\n";
				} else {
					(*pOut) << "<p>" << "Showing all results, which could have been split onto " << result.getPageCount() << " pages.</p>\n";
				}
			} else {
				BookcaseHitListConstIterator hitci = result.const_iterator();
				while (hitci.hasNext()) {
					BookcaseHit hit = hitci.next();
					(*pOut) << hit.m_bookcase_fm << '\t' << hit.m_bookcase_lm << '\t' << hit.m_hit_fm << '\n';
				}
				(*pOut) << std::flush;
			}
		}
			
		if (!bResult) {
			std::cerr << "ERROR doing harvesting:\n" 
			  << error_message
			  << std::endl;
		
		}
	}

	delete pFOUT;
	delete pFilterChain;

	if (bResult) {
		return 0;
	} else {
		std::cerr << "...Harvesting finished with error!" << std::endl;
		return 5;
	}
}

void show_version(std::ostream& ostr)
{
	ostr << "fts_harvester from Emdros version " << EMDROS_VERSION << std::endl;
}

void print_usage(std::ostream& ostr)
{
	show_version(ostr);
	ostr << "Usage:" << std::endl;
	ostr << "   fts_harvester [options] <query-terms>" << std::endl;
	ostr << "   fts_harvester (--help|-V|--version)" << std::endl;
	ostr << "OPTIONS:" << std::endl;
	printUsageStandardArguments(ostr);
	ostr << "   -d                   Set the database name to query (required)" << std::endl;
	ostr << "   -fe , --fts-engine (1|2|3)    Use the given FTS engine version" << std::endl;
	ostr << "   -g , --google-syntax If given, Google syntax will be used." << std::endl;
	ostr << "   --xhtml              If given, an XHTML snippet will be output." << std::endl;
	ostr << "   --bookcase-title-otn <object-type-name> The name of the OT for bc titles" << std::endl;
	ostr << "   -btf, --bookcase-title-feature <feature-name> The name of the feature" << std::endl;
	ostr << "                        from which to read the bookcase title. Only used" << std::endl;
	ostr << "                        with --xhtml." << std::endl;
	ostr << "   -sf , --stylesheet-filename <stylesheet-filename> The name of the file" << std::endl;
	ostr << "                        from which to read the JSON stylesheet." << std::endl;
	ostr << "   -s , --stylesheet <stylesheet-name> The name of the stylesheet to use" << std::endl;
	ostr << "                        from the stylesheet file." << std::endl;
	ostr << "   --bookcase-otn <otn> Use otn as the object type for the bookcase" << std::endl;
	ostr << "   --indexed-otn <otn>  Use otn as the object type to be indexed" << std::endl;
	ostr << "   --indexed-feature <fn>  Use fn as the feature name to index on\n"
	     << "                           the indexed object type to index." << std::endl;
ostr << "   -lp, --largest-proximity <integer>  Use this number as the largest\n"
	     << "                           proximity." << std::endl;
	ostr << "   -o , --output <filename>  Output to filename" << std::endl;
	ostr << "   -page <integer>            Select page (-1 selects all)" << std::endl;
	ostr << "   -hpp <positive-integer>    Max. hits per page" << std::endl;
	ostr << "DEFAULTS:\n";
	ostr << "   -fe 1" << std::endl;
	ostr << "   -lp " << MAX_MONAD << std::endl;
	ostr << "   -page -1 " << std::endl;
	ostr << "   -hpp 10 " << std::endl;

	printUsageDefaultsOfStandardArguments(ostr);
}

int main(int argc, char* argv[])
{
	// Set defaults
	eOutputKind output_kind = kOKConsole;
	eCharsets charset = kCSUTF8;
	eBackendKind source_backend_kind;
	std::string output_filename;
	std::string source_db_name;
	std::string source_hostname;
	std::string source_user;
	std::string source_password;
	std::string bookcase_otn;
	std::string indexed_otn;
	std::string indexed_feature;
	std::string stylesheet_file_name;
	std::string stylesheet_name;
	std::string bookcase_title_feature;
	std::string bookcase_title_otn;
	bool bUseGoogleSyntax = false;
	bool bSingleBookcaseHits = false;
	bool bGetXHTML = false;
	long largest_proximity;
	StringList query_string_list;
	int fts_engine_version = 1;
	int hits_per_page = 10;
	int page_number = -1;

	addStandardArguments(false // Do add -e | --encoding
			     );

	addOption("-fe", "--fts-engine", true,
		  "1",
		  "ERROR: -fe and --fts-engine must have a number (engine version) as their argument.\n"
		  "       example: -fe 1\n");

	addOption("-btf", "--bookcase-title-feature", true,
		  "",
		  "ERROR: -btf and --bookcase-title-feature must have a feature-name as their argument.\n"
		  "       example: -btf title\n");

	addOption("-btotn", "--bookcase-title-otn", true,
		  "",
		  "ERROR: -btotn and --bookcase-title-otn must have an object type name\n"
		  "       as their argument.\n"
		  "       example: -btotn enclosing_object_type\n");

	addOption("-lp", "--largest-proximity", true,
		  "",
		  "ERROR: -lp and --largest-proximity must have an integer as their argument.\n"
		  "       example: -lp 100\n");

	addOption("-hpp", "--hits-per-page", true,
		  "10",
		  "ERROR: -hpp and --hits-per-page must have an integer as their argument.\n"
		  "       example: -hpp 10\n");

	addOption("-page", "--page", true,
		  "-1",
		  "ERROR: -page and --page must have an integer as their argument.\n"
		  "       example: -page 0\n");

	addOption("-s", "--stylesheet", true,
		  "",
		  "ERROR: -s and --stylesheet must have a stylesheet name as their argument.\n"
		  "       example: -s 'englishstylesheet'\n");

	addOption("-sf", "--stylesheet-filename", true,
		  "",
		  "ERROR: -sf and --stylesheet-filename must have a filename as their argument.\n"
		  "       example: -sf /home/myuser/stylesheet.json\n");

	addOption("--bookcase-otn", "--bookcase-otn", true,
		  "",
		  "ERROR: --bookcase-otn must have an object type name as its argument.\n"
		  "       example: --bookcase-otn Document\n");
	addOption("--indexed-otn", "--indexed-otn", true,
		  "",
		  "ERROR: --indexed-otn must have an object type name as its argument.\n"
		  "       example: --indexed-otn Word\n");
	addOption("--indexed-feature", "--indexed-feature", true,
		  "",
		  "ERROR: --indexed-feature must have a feature name as its argument.\n"
		  "       example: --indexed-feature surface\n");
	addOption("-g", "--google-syntax", false);
	addOption("--single-bookcase-hits", "--single-bookcase-hits", false);
	addOption("--xhtml", "--xhtml", false);

	addOption("-d", "--dbname", true, 
		  "emdf",
		  "ERROR: -d and --dbname must have a database-name as their argument.\n"
		  "       example: -d book1\n");
	addOption("-o", "--output", true, 
		  "",
		  "ERROR: -o and --output must have a filename as their argument.\n"
		  "       example: -o indexcreation.mql\n");

	// Parse arguments
	std::list<std::string> surplus_arguments;

	std::string error_message;
	if (!parseArguments(argc, argv, error_message, surplus_arguments)) {
		print_usage(std::cerr);
		std::cerr << error_message << std::endl;
		return 1;
	} else {
		bool bShowVersion;
		bool bShowHelp;
		std::string error_message;
		if (!getStandardArguments(bShowVersion, bShowHelp,
					  source_hostname,
					  source_user,
					  source_password,
					  source_backend_kind,
					  charset,
					  error_message)) {
			print_usage(std::cerr);
			std::cerr << error_message << std::endl;
			return 1;
		}

		// -V | --version
		if (bShowVersion) {
			show_version(std::cout);
			return 0;
		}

		// --help
		if (bShowHelp) {
			print_usage(std::cout);
			return 0;
		}

		// largest proximity
		if (getArgumentPresent("-lp")) {
			std::string largest_proximity_str;
			getArgumentValue("-lp", largest_proximity_str);
			largest_proximity = string2long(largest_proximity_str);
		} else {
			largest_proximity = MAX_MONAD;
		}

		// hits per page
		if (getArgumentPresent("-hpp")) {
			std::string hits_per_page_str;
			getArgumentValue("-hpp", hits_per_page_str);
			hits_per_page = string2long(hits_per_page_str);
		} else {
			hits_per_page = 10;
		}

		// page number
		if (getArgumentPresent("-page")) {
			std::string page_number_str;
			getArgumentValue("-page", page_number_str);
			page_number = string2long(page_number_str);
		} else {
			page_number = -1;
		}


		// largest proximity
		if (getArgumentPresent("-lp")) {
			std::string largest_proximity_str;
			getArgumentValue("-lp", largest_proximity_str);
			largest_proximity = string2long(largest_proximity_str);
		} else {
			largest_proximity = MAX_MONAD;
		}

		// stylesheet name
		if (getArgumentPresent("-s")) {
			getArgumentValue("-s", stylesheet_name);
		} else {
			stylesheet_name = "";
		}

		// stylesheet filename
		if (getArgumentPresent("-sf")) {
			getArgumentValue("-sf", stylesheet_file_name);
		} else {
			stylesheet_file_name = "";
		}

		if ((stylesheet_name != "" &&
		     stylesheet_file_name == "")
		    || (stylesheet_name == "" &&
			stylesheet_file_name != "")) {
			print_usage(std::cerr);
			std::cerr << "The -s <stylesheet> and -sf <stylesheet-filename> options\nmust be given together!" << std::endl;
			return 1;
		}

		// bookcase title feature name
		if (getArgumentPresent("-btf")) {
			getArgumentValue("-btf", bookcase_title_feature);
		} else {
			bookcase_title_feature = "";
		}

		// bookcase title object type name
		if (getArgumentPresent("-btotn")) {
			getArgumentValue("-btotn", bookcase_title_otn);
		} else {
			bookcase_title_otn = "";
		}


		// source database name
		if (getArgumentPresent("-d")) {
			getArgumentValue("-d", source_db_name);
		} else {
			print_usage(std::cerr);
			std::cerr << "The -d <database-name> option is required!" << std::endl;
			return 1;
		}

		// bookcase-otn
		if (getArgumentPresent("--bookcase-otn")) {
			getArgumentValue("--bookcase-otn", bookcase_otn);
		} else {
			print_usage(std::cerr);
			std::cerr << "The --bookcase-otn <bookcase-otn> option is required!" << std::endl;
			return 1;
		}

		// indexed-otn
		if (getArgumentPresent("--indexed-otn")) {
			getArgumentValue("--indexed-otn", indexed_otn);
		} else {
			print_usage(std::cerr);
			std::cerr << "The --indexed-otn <indexed-otn> option is required!" << std::endl;
			return 1;
		}

		// indexed-feature
		if (getArgumentPresent("--indexed-feature")) {
			getArgumentValue("--indexed-feature", indexed_feature);
		} else {
			print_usage(std::cerr);
			std::cerr << "The --indexed-feature <indexed-feature> option is required!" << std::endl;
			return 1;
		}

		// output_filename
		if (getArgumentPresent("-o")) {
			getArgumentValue("-o", output_filename);
		} else {
			output_filename = "";
		}

		// fts engine version
		if (getArgumentPresent("-fe")) {
			std::string fts_engine_str;
			getArgumentValue("-fe", fts_engine_str);
			fts_engine_version = (int) string2long(fts_engine_str);
		} else {
			fts_engine_version = 1;
		}

		// -g, --google-syntax
		if (getArgumentPresent("-g")) {
			bUseGoogleSyntax = true;
		}

		// --single-bookcase-hits
		if (getArgumentPresent("--single-bookcase-hits")) {
			bSingleBookcaseHits = true;
		}

		// --xhtlm
		if (getArgumentPresent("--xhtml")) {
			bGetXHTML = true;
		}


		if (surplus_arguments.size() > 0) {
			std::list<std::string>::const_iterator ci = 
				surplus_arguments.begin();
			while (ci != surplus_arguments.end()) {
				query_string_list.addStringBack(*ci);
				++ci;
			}
		} 
	}

	if (!(fts_engine_version == 1 
	      || fts_engine_version == 2
	      || fts_engine_version == 3)) {
		std::string error_message = "ERROR: Requested FTS engine version " + long2string(fts_engine_version) + " is not available.\n"
			+ "Please use either 1, 2, or 3.";
		print_usage(std::cerr);
		std::cerr << error_message << std::endl;
		return 1;
	} 

	// Make connection

	EmdrosEnv *pSourceEE = new EmdrosEnv(output_kind, 
					     charset,
					     source_hostname,
					     source_user, source_password,
					     source_db_name,
					     source_backend_kind);

	// Zero-fill passwords
	zeroFillString(source_password);

	// Check that we connected to source
	if (!(pSourceEE->connectionOk())) {
		emdrosMessageConnectionNotOK(&std::cerr, source_backend_kind);
		delete pSourceEE;
		return 2;
	}

	int nResult = 3;
	try {
		nResult = exec_harvester(pSourceEE, output_filename, fts_engine_version, bookcase_otn, indexed_otn, indexed_feature, stylesheet_name, stylesheet_file_name, query_string_list, largest_proximity, bUseGoogleSyntax, bGetXHTML, bSingleBookcaseHits, bookcase_title_otn, bookcase_title_feature, hits_per_page, page_number);
	} catch (EMdFNULLValueException& e) {
		std::cerr << "ERROR: EMdFNULLValueException (Database error)..." << std::endl;
		std::cerr << pSourceEE->getDBError() << std::endl; 
		std::cerr << pSourceEE->getCompilerError() << std::endl;
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBDBError& e) {
		std::cerr << "ERROR: EMdFDBDBError (Database error)..." << std::endl;
		std::cerr << pSourceEE->getDBError() << std::endl; 
		std::cerr << pSourceEE->getCompilerError() << std::endl;
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFDBException& e) {
		std::cerr << "ERROR: EMdFDBException (Database error)..." << std::endl;
		std::cerr << pSourceEE->getDBError() << std::endl; 
		std::cerr << pSourceEE->getCompilerError() << std::endl;
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableColumnException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableRowException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (TableException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (BadMonadsException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (WrongCharacterSetException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EMdFOutputException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (EmdrosException& e) {
		std::cerr << e.what() << std::endl << "Program aborted." << std::endl;
	} catch (...) {
		std::cerr << "Unknown exception occurred.  Program aborted." << std::endl;
	} 

	// Clean up
	delete pSourceEE;

	return nResult;
}


